File: fizzbuzz.py
#this solution, however, breaks the solution in 15 lines, so it is not optimal 
for num in range(1,32):
  if num % 3 and num % 5 ==0:
    print('fizzbuzz')
  
  elif num % 5 ==0:
    print('buzz')

  elif num % 3 ==0:
    print('fizz')
  
  else:
    print(num)

#Solution 2 
# rearrange the order of the branches that print "Fizz" and "Buzz"
#"FizzBuzz" is never printed because, by definition, 
#any number that is eligible to be "FizzBuzz"'ed – divisible by 3 and 5 – 
# is also eligible to be either "Fizz" or "Buzz" – divisible by either 3 or 5.


for num in range(1, 32):
    if num % 5 is 0:
        print("Buzz")
    elif num % 3 is 0:
        print("Fizz")
    elif num % 15 is 0:
        print("FizzBuzz")
    else:
        print(num)

#In an if/elif/else construct, the interpreter will stop at the first branch that evaluates to True. 
# So the trick is to arrange the conditional branches so that the FizzBuzz branch takes priority over the other two.

for num in range(1,33): 
    if num % 15 is 0:
        print("FizzBuzz")
    elif num % 3 is 0:
        print("Fizz")
    elif num % 5 is 0:
        print("Buzz")
    else:
        print(num)

#solution 3 - function 

# import the pandas library
import pandas as pd
import numpy as np

# define the function
def fizz_buzz(num_sequence_f):
    if (num_sequence_f % 3 == 0) & (num_sequence_f % 5 == 0):
        return('FizzBuzz')
    elif (num_sequence_f % 3 == 0):
        return('Fizz')
    elif (num_sequence_f % 5 == 0):
        return('Buzz')
    else:
        return(num_sequence_f)

# apply it to the number series
pd.Series(np.arange(1,33)).apply(fizz_buzz)

# numpy arange command (which returns an array of numbers).
# As described above, we convert the array to a pandas Series, which is then passed to our function.

File: fizzbuzz.r
# R has the standard control structures you would expect. 
# expr can be multiple (compound) statements by enclosing them in braces { }.
# It is more efficient to use built-in functions rather than control structures whenever possible

#SOLUTION 1  - for loop 

#  We start with the double evaluation – 
# any other ordering of our if-else statements would cause the program to fail 
# (because numbers which pass the double evaluation by definition also pass the single evaluations).
 
for (i in 1:32){
 if(i%%3 == 0 & i%%5 == 0) {
    print('FizzBuzz')
  }
  else if(i%%3 == 0) {
    print('Fizz')
  }
  else if (i%%5 == 0){
    print('Buzz')
  }
  else {
    print(i)
  }
  
}

#solution 2 - use function 
#A for loop conveys that it’s iterating over something, but doesn’t clearly convey a high level goal
# defines a function called “fizz_buzz”. This function takes an input (which I have defined in the function as num_sequence_f), 
# and then passes it through the logical sequence we defined using the “if” statements above.

fizz_buzz <- function(num_sequence_f){
  if(num_sequence_f%%3 == 0 & num_sequence_f%%5 == 0) {
    print('FizzBuzz')
  }
  else if(num_sequence_f%%3 == 0) {
    print('Fizz')
  }
  else if (num_sequence_f%%5 == 0){
    print('Buzz')
  }
  else {
    print(num_sequence_f)
  }
  
}


# apply it to the numbers 1 to 100
sapply(seq(from = 1, to = 100, by = 1), fizz_buzz)

